src/systems/tunnel-effect.js

```diff
@@ -0,0 +1,164 @@
+import "../utils/postprocessing/EffectComposer";
+import "../utils/postprocessing/RenderPass";
+import "../utils/postprocessing/ShaderPass";
+import "../utils/postprocessing/MaskPass";
+import "../utils/shaders/CopyShader";
+import "../utils/shaders/VignetteShader";
+import qsTruthy from "../utils/qs_truthy";
+
+const isDisabled = qsTruthy("disableTunnel");
+const CLAMP_SPEED = 0.01;
+const CLAMP_RADIUS = 0.001;
+const CLAMP_SOFTNESS = 0.001;
+const FADE_TIMEOUT_MS = 300;
+const TARGET_RADIUS = 10.0;
+const TARGET_SOFTNESS = 0.0;
+
+AFRAME.registerSystem("tunneleffect", {
+  schema: {
+    targetComponent: { type: "string", default: "character-controller" },
+    radius: { type: "number", default: 1.0, min: 0.25 },
+    minRadius: { type: "number", default: 0.25, min: 0.1 },
+    maxSpeed: { type: "number", default: 0.7, min: 0.1 },
+    softness: { type: "number", default: 0.1, min: 0.0 },
+    opacity: { type: "number", default: 1, min: 0.0 }
+  },
+
+  init: function() {
+    this.scene = this.el;
+    this.isMoving = false;
+    this.isVR = false;
+    this.dt = 0;
+    this.isPostProcessingReady = false;
+    this.characterEl = document.querySelector(`a-entity[${this.data.targetComponent}]`);
+    if (this.characterEl) {
+      this._initPostProcessing = this._initPostProcessing.bind(this);
+      this.characterEl.addEventListener("componentinitialized", this._initPostProcessing);
+    } else {
+      console.warn("Could not find target component.");
+    }
+    this._enterVR = this._enterVR.bind(this);
+    this._exitVR = this._exitVR.bind(this);
+    this.scene.addEventListener("enter-vr", this._enterVR);
+    this.scene.addEventListener("exit-vr", this._exitVR);
+  },
+
+  pause: function() {
+    if (!this.characterEl) {
+      return;
+    }
+    this.characterEl.removeEventListener("componentinitialized", this._initPostProcessing);
+    this.scene.removeEventListener("enter-vr", this._enterVR);
+    this.scene.removeEventListener("exit-vr", this._exitVR);
+  },
+
+  play: function() {
+    this.scene.addEventListener("enter-vr", this._enterVR);
+    this.scene.addEventListener("exit-vr", this._exitVR);
+  },
+
+  tick: function(t, dt) {
+    this.dt = dt;
+
+    if (isDisabled || !this.isPostProcessingReady || !this.isVR) {
+      return;
+    }
+
+    const characterSpeed = this.characterComponent.velocity.length();
+    if (characterSpeed < CLAMP_SPEED) {
+      // the character stops, so we use the aframe default render func
+      const r = this.vignettePass.uniforms["radius"].value;
+      const softness = this.vignettePass.uniforms["softness"].value;
+      if (
+        this.isMoving &&
+        Math.abs(r - TARGET_RADIUS) > CLAMP_RADIUS &&
+        Math.abs(softness - TARGET_SOFTNESS) > CLAMP_SOFTNESS
+      ) {
+        if (this.deltaR !== 0 && this.deltaS !== 0) {
```

Comment:
I think this line will only be true once, when `this.deltaR` and `this.deltaS` are `undefined.`  Is that desired?